### PR TITLE
[EDA]: Prevent Deletion of Course Connection if There are Associated Child Records

### LINKED ISSUE #1046: [EDA] Prevent Deletion of Course Connection If It Has Related Child Records
**Note: FOR INTERNAL USE ONLY. For support questions, please use the EDA Chatter group in [The Power of Us Hub](https://powerofus.force.com).
For feature requests, please use the Ideas tab in [The Power of Us Hub](https://powerofus.force.com).**

* **Do you want to request a *feature* or report a *bug*?**
Feature 
* **What is the current behavior?**
When users delete a Course Connection with child records, the system will allow users to delete the Course Connection and it will not delete child records. 
* **What is the expected behavior?**
We want to provide customers the ability to prevent deletion of Course Connection if it has related child records by allowing them to toggle a checkbox in the Hierarchy Setttings. 
* **What is the motivation / use case for changing the behavior?**
By giving customers the ability to control whether a Course Connection with child records should be deleted or not, it will help them with maintaining a clean system. 
* **Acceptance Criteria**

  - **Given**: If Prevent_Course_Connection_Deletion__c checkbox is enabled in Hierarchy Settings
  - **When**: I try to delete a Course Connection record with child records
  - **Then**: I should not be able to delete the Course Connection record.

### PR DESCRIPTION
**Pull request by kdy618**
# Critical Changes

# Changes
## Updates to Ensure Data Integrity
We've added functionality to prevent the deletion of Course Connection records that have related Term Grade or Attendance Event records. This change is for internal-use only and is disabled at this time. You won't see any difference in functionality.
**Note:** We strongly recommend that EDA Admins don't manage Hierarchy Custom Settings directly and, instead, use the EDA Settings UI.

# Issues Closed
Closes #1046
# New Metadata
- CCON_CannotDelete_TDTM
- CCON_CannotDelete_TEST

# Deleted Metadata

# Testing Notes

* Scenario 1: “Prevent Course Connection Deletion” checkbox is enabled and Course Connection record has related Term Grade or Attendance Event records. 
    * Check “Prevent Course Connection Deletion” checkbox in Hierarchy Settings. 
    * Create a Contact with a Course Connection. 
    * Add a Term Grade or Attendance Event or both to the Course Connection record. 
    * Delete the Course Connection record. 
    * You should receive a pop up error that you cannot delete this Course Connection. See https://quip.com/FUGPAUO0JSbY#UBKACA9I3wO for the error message. 

* Scenario 2: “Prevent Course Connection Deletion” checkbox is enabled and Course Connection record does not have related Term Grade or Attendance Event records. 
    * Check “Prevent Course Connection Deletion” checkbox in Hierarchy Settings. 
    * Create a Contact with a Course Connection. 
    * Delete the Course Connection record. 
    * You should be able to delete Course Connection record. 

* Scenario 3: “Prevent Course Connection Deletion” checkbox is disabled and Course Connection record has related Term Grade or Attendance Event records. 
    * Check “Prevent Course Connection Deletion” checkbox in Hierarchy Settings. 
    * Create a Contact with a Course Connection. 
    * Add a Term Grade or Attendance Event or both to the Course Connection record. 
    * Delete the Course Connection record. 
    * You should be able to delete Course Connection record. 

* Scenario 4: “Prevent Course Connection Deletion” checkbox is disabled and Course Connection record does not have related Term Grade or Attendance Event records. 
    * Check “Prevent Course Connection Deletion” checkbox in Hierarchy Settings. 
    * Create a Contact with a Course Connection. 
    * Delete the Course Connection record. 
    * You should be able to delete Course Connection record.

### CODE REVIEW

File: src/classes/CCON_CannotDelete_TDTM.cls [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+     TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+
+        Map<ID, SObject> oldmap = new Map<ID, SObject>(oldlist);
+        Boolean isCourseConnectionDeletionEnabled = UTIL_CustomSettingsFacade.getSettings().Prevent_Course_Connection_Deletion__c;
+
+        if (triggerAction == TDTM_Runnable.Action.BeforeDelete && isCourseConnectionDeletionEnabled) {
+            for(Course_Enrollment__c ce : [SELECT Id,
+                                          (SELECT Id FROM Term_Grades__r LIMIT 1),
+                                          (SELECT Id FROM Attendance_Events__r LIMIT 1)
+                                          FROM Course_Enrollment__c WHERE Id in: oldlist]) {
```
Comment (skosgi1) [inline src/classes/CCON_CannotDelete_TDTM.cls]: Minor IN:

File: src/classes/CCON_CannotDelete_TEST.cls [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        Account acc = UTIL_UnitTestData_TEST.getMultipleTestAccounts(1, UTIL_Describe_API.getBizAccRecTypeID())[0];
+        insert acc;
+
+        Contact[] contacts = UTIL_UnitTestData_TEST.getMultipleTestContacts(6);
+        insert contacts;
+
+        Course_Offering__c courseOffering = UTIL_UnitTestData_TEST.createCourseOffering(NULL, NULL);
+
+        List<Course_Enrollment__c> courseConnections = new List<Course_Enrollment__c>{
+            new Course_Enrollment__c(Contact__c = contacts[0].Id, Course_Offering__c = courseOffering.Id),
```
Comment (skosgi1) [inline src/classes/CCON_CannotDelete_TEST.cls]: Also, can we use getCourseConnection method in UTIL_UnitTestdata_TEST to create test records here? Something like

Course_Enrollment__c courseConnection1 = UTIL_UnitTestdata_TEST. getCourseConnection(contacts[0].Id, courseOffering.Id);

Course_Enrollment__c courseConnection2 = 
insert new Course_Enrollment__c [] {courseConnection1, courseConnection2};

File: src/classes/CCON_CannotDelete_TEST.cls [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            new Course_Enrollment__c(Contact__c = contacts[1].Id, Course_Offering__c = courseOffering.Id),
+            new Course_Enrollment__c(Contact__c = contacts[2].Id, Course_Offering__c = courseOffering.Id),
+            new Course_Enrollment__c(Contact__c = contacts[3].Id, Course_Offering__c = courseOffering.Id),
+            new Course_Enrollment__c(Contact__c = contacts[4].Id, Course_Offering__c = courseOffering.Id)};
+        insert courseConnections;
+
+        List<Term_Grade__c> termGrades = new List<Term_Grade__c>{
+            new Term_Grade__c(Contact__c = contacts[0].Id, Course_Offering__c = courseOffering.Id,
```
Comment (skosgi1) [inline src/classes/CCON_CannotDelete_TEST.cls]: Can we do the same with term grades using getTermGrade method

File: src/classes/CCON_CannotDelete_TEST.cls [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                                Course_Connection__c = courseConnections[0].Id),
+            new Term_Grade__c(Contact__c = contacts[2].Id, Course_Offering__c = courseOffering.Id,
+                                Course_Connection__c = courseConnections[2].Id)
+        };
+
+        List<Attendance_Event__c> attendanceEvents = new List<Attendance_Event__c>{
+		    new Attendance_Event__c(Contact__c = contacts[1].Id, Date__c = System.today(),
```
Comment (skosgi1) [inline src/classes/CCON_CannotDelete_TEST.cls]: same here getAttendanceEvent ?

File: src/classes/CCON_CannotDelete_TEST.cls [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        Term_Grade__c termGrade1 = UTIL_UnitTestData_TEST.getTermGrade(contacts[0].Id, courseOffering.Id, courseConnections[0].Id);
+        Term_Grade__c termGrade2 = UTIL_UnitTestData_TEST.getTermGrade(contacts[2].Id, courseOffering.Id, courseConnections[2].Id);
+        termGrades.add(termGrade1);
+        termGrades.add(termGrade2);
+
+        List<Attendance_Event__c> attendanceEvents = new List<Attendance_Event__c>();
+        Attendance_Event__c attendanceEvent1 = UTIL_UnitTestData_TEST.getAttendanceEvent();
+        attendanceEvent1.Contact__c = contacts[1].Id;
+        attendanceEvent1.Course_Connection__c = courseConnections[1].Id;
+        Attendance_Event__c attendanceEvent2 = UTIL_UnitTestData_TEST.getAttendanceEvent();
```
Comment (skosgi1) [inline src/classes/CCON_CannotDelete_TEST.cls]: Oh looks like we have another getAttendanceEvent() method with 4 parameters can you use that instead? would reduce a few lines of code here.

getAttendanceEvent(Id contactId, Id courseConnectionID, String attendanceType, String reason)

File: src/classes/CCON_CannotDelete_TEST.cls [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Course Enrollments
+* @group-content ../../ApexDocContent/CourseEnrollments.htm
+* @description Tests for CCON_CannotDelete_TDTM.
+*/
+@isTest
+private with sharing class CCON_CannotDelete_TEST {
+    @isTest
```
Comment (skosgi1) [inline src/classes/CCON_CannotDelete_TEST.cls]: Sorry I missed this earlier, can you please write a brief description of the test method does as comment before the method starts? something like
/******************************************************
* @description
*
*************************************************************/

File: src/classes/CCON_CannotDelete_TEST.cls [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        //Verifies Course Connection with Attendance Event was not deleted
+        System.assertEquals(false, results[1].success);
+        System.assertEquals(Label.CannotDelete, results[1].errors[0].message);
+
+        //Verifies Course Connection with Attendance Event and Term Grade was not deleted.
+        System.assertEquals(false, results[2].success);
+        System.assertEquals(Label.CannotDelete, results[2].errors[0].message);
+
+    }
+
```
Comment (skosgi1) [inline src/classes/CCON_CannotDelete_TEST.cls]: same here

File: src/classes/CCON_CannotDelete_TEST.cls [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        termGrades.add(termGrade1);
+        termGrades.add(termGrade2);
+
+        List<Attendance_Event__c> attendanceEvents = new List<Attendance_Event__c>();
+        Attendance_Event__c attendanceEvent1 = UTIL_UnitTestData_TEST.getAttendanceEvent();
+        attendanceEvent1.Contact__c = contacts[1].Id;
+        attendanceEvent1.Course_Connection__c = courseConnections[1].Id;
+        Attendance_Event__c attendanceEvent2 = UTIL_UnitTestData_TEST.getAttendanceEvent();
+        attendanceEvent2.Contact__c = contacts[2].Id;
+        attendanceEvent2.Course_Connection__c = courseConnections[2].Id;
```
[the remark this answers is not in the document]
Comment (skosgi1) [reply src/classes/CCON_CannotDelete_TEST.cls]: This was regarding https://github.com/SalesforceFoundation/EDA/pull/1042#discussion_r364486032 below. Sry for the confusion.

File: src/classes/CCON_CannotDelete_TEST.cls (hunk 7 above)
Comment (skosgi1) [reply src/classes/CCON_CannotDelete_TEST.cls]: Sry, I meant @description for the 2nd test method.

File: src/classes/CCON_CannotDelete_TEST.cls (hunk 5 above)
Comment (kdy618) [reply src/classes/CCON_CannotDelete_TEST.cls]: @skosgi1 , do you know why we have 2 of the same method one without the signature, and one with?

Comment (skosgi1) [reply src/classes/CCON_CannotDelete_TEST.cls]: I think the first one was an old one...the second one was created to pass a few paramters I think

**Comment by spinkelman** [comment]
Passes QA | Merging
Closing WI

**Pull request merged**